node/node.go

```diff
@@ -884,27 +905,24 @@ func initApplicationRetainHeight(stateStore sm.Store) error {
 	return nil
 }
 
-func initCompanionBlockRetainHeight(stateStore sm.Store, companionEnabled bool, initialRetainHeight int64) error {
-	if _, err := stateStore.GetCompanionBlockRetainHeight(); err != nil {
-		// If the data companion block retain height has not yet been set in
-		// the database
-		if errors.Is(err, sm.ErrKeyNotFound) {
-			if companionEnabled && initialRetainHeight > 0 {
-				// This will set the data companion retain height into the
-				// database. We bypass the sanity checks by
-				// pruner.SetCompanionBlockRetainHeight. These checks do not
-				// allow a retain height below the current blockstore height or
-				// above the blockstore height  to be set. But this is a retain
-				// height that can be set before the chain starts to indicate
-				// potentially that no pruning should be done before the data
-				// companion comes online.
-				err = stateStore.SaveCompanionBlockRetainHeight(initialRetainHeight)
-				if err != nil {
-					return fmt.Errorf("failed to set initial data companion block retain height: %w", err)
-				}
-			}
-		} else {
-			return fmt.Errorf("failed to obtain companion retain height: %w", err)
+// Sets the data companion retain heights if one of two possible conditions is met:
+// 1. One or more of the retain heights has not yet been set.
+// 2. One or more of the retain heights is currently 0.
+func initCompanionRetainHeights(stateStore sm.Store, initBlockRH, initBlockResultsRH int64) error {
+	if curBlockRH, err := stateStore.GetCompanionBlockRetainHeight(); err != nil || curBlockRH == 0 {
+		if !errors.Is(err, sm.ErrKeyNotFound) {
```

Comment:
this logic is triggering an error in my tests, the `GetCompanionBlockRetainHeight` is returning `err = nil` and `curBlockRH = 0` (because the key is not set yet) and in the test to check if `err` is not `sm.ErrKeyNotFound` it is true (`nil` != `ErrKeyNotFound`) and triggered, and the node fails to start with a log message:

 `ERROR: failed to create node: failed to create pruner: failed to obtain companion block retain height: %!w(<nil>)`

Reply:
also, to rely on `curBlockRH == 0` to assume the value is not set might cause some undesired behaviors I believe, since in the `GetCompanionBlockRetainHeight` there's a test

```    buf, err := store.getValue(CompanionBlockRetainHeightKey)
	if err != nil {
		return 0, err
	}
```
I believe there's a possibility that if you get an error in `store.getValue()` (e.g. disk issue, etc) that might return 0, and this logic might reset the current retain height value stored to the default config value. 

For example, let's say current retain height was set to `100` by the Data Companion, then a node is stopped and re-started. There might be a possibility it will get reset to the default value in the config if there's an error in that function call. I don't think this might cause major issues because probably the default config value will be a lower value, but just mentioning here.

Reply:
Yeah this is what's causing the E2E failure. I've fixed it locally now, but that's causing other tests to fail. Once I'm done playing whack-a-mole with the test failures I'll update :slightly_smiling_face: